Add common test logging methods such as Errorf/Fatalf/Printf

13rac1 opened the pull request:
Implements nearly all of the test logging methods for both T and B
structs. Majority of the code has been copied from:
golang.org/src/testing/testing.go
then updated to match the existing testing.go structure.

Code structure/function/method order mimics upstream.

Both FailNow() and SkipNow() cannot be completely implemented,
because they require an early exit from the goroutine. Instead,
they call Error() to report the limitation.

This incomplete implementation allows more detailed test logging and
increases compatiblity with upstream.

aykevl commented:
This is a very useful addition. However, I'm slightly worried about simply copying the documentation strings from upstream without attribution. Can you add a comment to the top of testing.go that indicates where they came from, something like this?

```go
// Some API comments have been copied from the main Go implementation and
// copyright by the Go Authors. See the following license for details:
//     https://golang.org/LICENSE
```

@deadprogram this isn't my area of expertise, do you have an idea how to best handle this?

deadprogram commented:
BSD allows for this sort of thing as long as the BSD license notice is included if redistributed in source code format.

aykevl commented:
Does that mean the PR can be merged as-is?

13rac1 commented:
I was wondering the same thing, but figured you already had some upstream code and had a process for re-using it.

FYI I've rebased vs `dev` and added `common` to the `B` struct for all logging funcs to work.

13rac1 commented:
Does this need anything else to be merged? Do you want me to add the sourcing comment? The wasm_exec.js has this: https://github.com/tinygo-org/tinygo/blob/7d5542dda73e8dda066b604870eac279cd644a7e/targets/wasm_exec.js#L1-L5

deadprogram commented:
Yes, please add the sourcing comment, to make it easier for the reader to determine the provenance of the code. Also, if we need to update it, that way we know where to look.

Thanks!

13rac1 commented:
Source license, file URL with SHA, and some additional func doc strings added.

13rac1 commented on src/testing/testing.go:
```diff
@@ -7,13 +15,143 @@ import (
+	SkipNow()
+	Skipf(format string, args ...interface{})
+	Skipped() bool
+	// Helper()
```
I've removed this on purpose, but it could be added as NO OP too.

deadprogram commented:
Looking good now, thanks for the contribution @13rac1 now merging.
